samples/net/prometheus/src/main.c

```diff
@@ -0,0 +1,222 @@
+#include <zephyr/kernel.h>
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/net/http/server.h>
+#include <zephyr/net/http/service.h>
+#include <zephyr/net/net_ip.h>
+#include <zephyr/net/socket.h>
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/counter.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_DBG);
+
+struct {
+
+	struct prometheus_collector *collector;
+
+	struct prometheus_counter *counter;
+
+} prom_context;
+
+#if defined(CONFIG_NET_SAMPLE_HTTP_SERVICE)
+static uint16_t test_http_service_port = CONFIG_NET_SAMPLE_HTTP_SERVER_SERVICE_PORT;
+HTTP_SERVICE_DEFINE(test_http_service, CONFIG_NET_CONFIG_MY_IPV4_ADDR, &test_http_service_port, 1,
+		    10, NULL);
+
+static uint8_t recv_buffer[2048];
+
+static int dyn_handler(struct http_client_ctx *client, enum http_data_status status,
+		       uint8_t *buffer, size_t len, void *user_data)
+{
+#define MAX_TEMP_PRINT_LEN 32
+	int rc;
+	char hex_len[MAX_TEMP_PRINT_LEN];
+	static char print_str[MAX_TEMP_PRINT_LEN];
+	enum http_method method = client->method;
+	static size_t processed;
+	static bool replied;
+
+	__ASSERT_NO_MSG(buffer != NULL);
+
+	if (status == HTTP_SERVER_DATA_ABORTED) {
+		LOG_DBG("Transaction aborted after %zd bytes.", processed);
+		processed = 0;
+		replied = false;
+		return 0;
+	}
+
+	processed += len;
+
+	snprintf(print_str, sizeof(print_str), "%s received (%zd bytes)", http_method_str(method),
+		 len);
+	LOG_HEXDUMP_DBG(buffer, len, print_str);
+
+	if (status == HTTP_SERVER_DATA_FINAL) {
+		LOG_DBG("All data received (%zd bytes).", processed);
+		processed = 0;
+	}
+
+	if (strlen(buffer) == 0) {
+		replied = false;
+	}
+
+	/* incrase counter per request */
+	prometheus_counter_inc(prom_context.counter);
+
+	(void)memset(recv_buffer, 0, sizeof(recv_buffer));
+
+	prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
+
+	/**
+	 * The response status line indicates that the request was successful (200 OK).
+	 * The response headers include Date, Server, Content-Type, and Transfer-Encoding.
+	 * Transfer-Encoding is set to chunked, indicating that the response body is sent in a
+	 * series of chunks. Each chunk begins with the hexadecimal length of the chunk, followed by
+	 * a carriage return and line feed ("\r\n"), then the chunk data itself, and finally
+	 * terminated by another carriage return and line feed ("\r\n"). The last chunk has a length
+	 * of 0, indicating the end of the response body.
+	 *
+	 */
+
+	rc = snprintk(hex_len, sizeof(hex_len), "%x\r\n", strlen(recv_buffer));
+	if (rc < 0) {
+		LOG_ERR("Cannot format data length (%d)", rc);
+		return rc;
+	}
+
+	memcpy(recv_buffer + strlen(recv_buffer), "\r\n", 2);
```

Comment:
What happens if `strlen(recv_buffer) == sizeof(recv_buffer)` here?
One more point for [`net_buf`](https://docs.zephyrproject.org/latest/services/net_buf/index.html)

Reply:
That being said, it is also possible that `recv_buffer` is large enough for anything coming from here: https://github.com/zephyrproject-rtos/zephyr/pull/72804/files/7c6366cd4d818e68a6f18cf31a108ac979dc74c9#diff-39b74f8d5670ba8f827d9fbc421275e870d519da3b2121771edb00d6e7157c27R85